File: composeApp/src/androidMain/kotlin/io/github/a7asoft/ui/components/ErrorPanel.kt
package io.github.a7asoft.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.github.a7asoft.nfcemv.reader.IoReason
import io.github.a7asoft.nfcemv.reader.ReaderError

/**
 * Renders a [ReaderError] as a friendly headline plus a collapsible
 * diagnostic detail (`error.toString()`) and a "Try again" CTA.
 *
 * `error.toString()` is safe to display: every [ReaderError] variant
 * carries only structural metadata (status words, structural sub-error
 * references, [IoReason] enum) — never raw card bytes or arbitrary
 * exception messages.
 */
@Composable
public fun ErrorPanel(error: ReaderError, onTryAgain: () -> Unit) {
    val expandedState = remember { mutableStateOf(false) }
    val expanded = expandedState.value
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.errorContainer),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text(text = friendlyMessage(error), style = MaterialTheme.typography.titleMedium)
            OutlinedButton(onClick = { expandedState.value = !expandedState.value }) {
                Text(if (expanded) "Hide details" else "Show details")
            }
            if (expanded) {
                Text(text = error.toString(), style = MaterialTheme.typography.bodySmall)
            }
            OutlinedButton(onClick = onTryAgain, modifier = Modifier.fillMaxWidth()) {
                Text("Try again")
            }
        }
    }
}

@Suppress("CyclomaticComplexMethod")
// why: exhaustive when over sealed [ReaderError] catalogue. Each branch is
// a single mapping, not real complexity.
private fun friendlyMessage(error: ReaderError): String = when (error) {
    is ReaderError.IoFailure -> friendlyIoMessage(error.reason)
    ReaderError.PpseUnsupported -> "This card does not advertise a contactless EMV directory."
    ReaderError.NoApplicationSelected -> "The card listed no contactless applications."
    is ReaderError.ApduStatusError -> "The card returned an unexpected status."
    is ReaderError.PpseRejected -> "The PPSE response could not be parsed."
    is ReaderError.GpoRejected -> "The GET PROCESSING OPTIONS response could not be parsed."
    is ReaderError.ParseFailed -> "The card data could not be parsed into an EMV card."
}

@Suppress("CyclomaticComplexMethod")
// why: 3-branch sealed dispatch over [IoReason]; detekt counts the
// function itself as +1, tipping the threshold. The suppression matches
// that algorithm.
private fun friendlyIoMessage(reason: IoReason): String = when (reason) {
    IoReason.TagLost -> "The card was moved away too quickly. Hold it steady."
    IoReason.Timeout -> "The card took too long to respond. Try again."
    IoReason.Generic -> "Communication with the card failed."
}

File: iosApp/iosApp/ui/components/ErrorPanel.swift
import SwiftUI
import EmvReader

/// Renders a ``ReaderError`` as a friendly headline plus a collapsible
/// diagnostic detail and a "Try again" CTA.
///
/// The `friendlyMessage` and `diagnostic` strings are pre-computed by
/// ``ReaderViewModel`` (via `ErrorMessages`) so this view stays a
/// pure renderer. Diagnostics are safe to display: every
/// ``ReaderError`` variant carries only structural metadata (status
/// words, structural sub-error references, ``IoReason`` enum) — never
/// raw card bytes (CLAUDE.md §5.8).
public struct ErrorPanel: View {

    private let friendlyMessage: String
    private let diagnostic: String
    private let onTryAgain: () -> Void
    @State private var detailsExpanded: Bool = false

    /// Build an error panel. The caller passes pre-rendered messages
    /// so the view is a pure function of its inputs.
    public init(
        friendlyMessage: String,
        diagnostic: String,
        onTryAgain: @escaping () -> Void
    ) {
        self.friendlyMessage = friendlyMessage
        self.diagnostic = diagnostic
        self.onTryAgain = onTryAgain
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(friendlyMessage)
                .font(.headline)
            DisclosureGroup("Details", isExpanded: $detailsExpanded) {
                Text(diagnostic)
                    .font(.system(.caption, design: .monospaced))
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.top, 4)
            }
            Button("Try again", action: onTryAgain)
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)
        }
        .padding()
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color(.systemRed).opacity(0.1))
        )
    }
}
